packages/artillery-plugin-publish-metrics/lib/open-telemetry/index.js

```diff
@@ -0,0 +1,172 @@
+          config.serviceName || 'Artillery-test'
+      })
+    );
+
+    if (config.traces) {
+      this.tracing = true;
+      this.traceConfig = config.traces;
+
+      this.configureTrace(this.traceConfig);
+
+      attachScenarioHooks(script, [
+        {
+          type: 'beforeRequest',
+          name: 'startOTelSpan',
+          hook: this.startOTelSpan.bind(this)
+        }
+      ]);
+
+      attachScenarioHooks(script, [
+        {
+          type: 'afterResponse',
+          name: 'exportOTelSpan',
+          hook: this.exportOTelSpan.bind(this)
+        }
+      ]);
+    }
+  }
+
+  configureTrace(config) {
+    debug('Configuring Tracer Provider');
+    const {
+      BasicTracerProvider,
+      TraceIdRatioBasedSampler,
+      ParentBasedSampler,
+      BatchSpanProcessor
+    } = require('@opentelemetry/sdk-trace-base');
+
+    this.tracerOpts = {
+      resource: this.resource
+    };
+    if (config.sampleRate) {
+      this.tracerOpts.sampler = new ParentBasedSampler({
+        root: new TraceIdRatioBasedSampler(config.sampleRate)
+      });
+    }
+
+    this.tracerProvider = new BasicTracerProvider(this.tracerOpts);
+
+    debug('Configuring Exporter');
+    this.traceExporterOpts = {};
+    if (config.endpoint) {
+      this.traceExporterOpts.url = config.endpoint;
+    }
+
+    if (config.headers) {
+      this.traceExporterOpts.headers = config.headers;
+    }
+
+    const protocol = !config.protocol
+      ? 'http'
+      : this.protocols[config.protocol];
+
+    const {
+      OTLPTraceExporter
+    } = require(`@opentelemetry/exporter-trace-otlp-${protocol}`);
+    this.exporter = new OTLPTraceExporter(this.traceExporterOpts);
+
+    this.tracerProvider.addSpanProcessor(
+      new BatchSpanProcessor(this.exporter, {
+        scheduledDelayMillis: 1000
+      })
+    );
+    this.tracerProvider.register();
+  }
+
+  startOTelSpan(req, userContext, events, done) {
+    debug('Starting span');
+    const startTime = Date.now();
+    userContext.vars['__otlStartTime'] = startTime;
+    const span = trace
+      .getTracer('artillery-tracer')
+      .startSpan('http_request', { startTime });
+    span.setAttribute('kind', 'client');
+    span.addEvent('http_request_started', startTime);
+    userContext.vars['__otlpSpan'] = span;
+
+    return done();
+  }
+
+  exportOTelSpan(req, res, userContext, events, done) {
+    if (!userContext.vars['__otlpSpan']) {
+      return done();
+    }
+
+    const span = userContext.vars['__otlpSpan'];
+    let endTime;
+
+    if (res.timings && res.timings.phases) {
+      span.setAttribute('responseTimeMs', res.timings.phases.firstByte);
+      endTime =
+        userContext.vars['__otlStartTime'] + res.timings.phases.firstByte;
+      span.addEvent('http_request_ended', endTime);
+    } else {
+      span.addEvent('http_request_ended');
+    }
+
+    const url = new URL(req.url);
+    span.setAttributes({
+      url: url.href,
+      'url.host': url.host,
```

Comment:
Should we send `server.address` and `server.port` instead of `url.host` value here to follow the conventions here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md

Reply:
Ah yes, good call, I am on it. 🙌🏻